src/v/pandaproxy/schema_registry/avro.cc

```diff
@@ -16,12 +16,55 @@
 #include <avro/Compiler.hh>
 #include <avro/Exception.hh>
 #include <avro/GenericDatum.hh>
+#include <avro/Types.hh>
 #include <avro/ValidSchema.hh>
 
 namespace pandaproxy::schema_registry {
 
 namespace {
 
+bool can_promote(avro::Node& writer, avro::Node& reader) {
+    switch (writer.type()) {
+    case avro::AVRO_INT: {
+        switch (reader.type()) {
+        case avro::AVRO_LONG:
+        case avro::AVRO_FLOAT:
+        case avro::AVRO_DOUBLE:
+            return true;
+        default:
+            return false;
+        }
+        return false;
+    }
+    case avro::AVRO_LONG: {
+        switch (reader.type()) {
+        case avro::AVRO_FLOAT:
+        case avro::AVRO_DOUBLE:
+            return true;
+        default:
+            return false;
+        }
+        return false;
+    }
+    case avro::AVRO_FLOAT: {
+        switch (reader.type()) {
+        case avro::AVRO_DOUBLE:
+            return true;
+        default:
+            return false;
+        }
+        return false;
+    }
+    case avro::AVRO_STRING:
+        return reader.type() == avro::AVRO_BYTES;
+    case avro::AVRO_BYTES:
+        return reader.type() == avro::AVRO_STRING;
```

Comment:
not knowing anything about Avro, why does AVRO_STRING match to an equality check against BYTES (and the opposite for AVRO_BYTES)?

Reply:
Because the spec says so:

Reply:
Ahhhh makes sense!